11538 poly solve issue

ctefer opened the pull request:
Closes #11538

Updated constructor to treat NumberSymbols as symbols. polys containing non-algebraic number symbols are now treated as ZZ. This allows solvers to solve polynomials containing these specialized constants.

Tests created were checked against WA. A test that ensures an empty set is not produced when given the original issue parameters was also produced. The non-empty set is hard to check against WA and contains tuples that need to be solved.

A test was updated for test_solveset because the solver method now simplifies the output even further than was originally asserted.

ctefer committed:
Commiting my initial mess up

ctefer committed:
Updated test for comment

jksuom commented on sympy/polys/constructor.py:
```diff
@@ -122,8 +123,12 @@ def _construct_composite(coeffs, opt):
+            if gen.is_number:
+                if isinstance(gen, NumberSymbol) and not gen.is_algebraic:
+                    continue
+                # generators are number-like so lets better use EX
```
It is probably not necessary to test for NumberSymbols explicitly. That could possibly give
```
    if any(gen.is_number and gen.is_algebraic for gen in gens):
        return None
```

jksuom commented on sympy/polys/tests/test_constructor.py:
```diff
@@ -121,3 +121,18 @@ def test_precision():
+
+def test_issue_11538():
+    dom = ZZ[E]
+    assert construct_domain(E) == (dom, dom.convert(E))
```
We may omit the `dom.convert` parts that are not of interest here. These two lines could be combined as follows:
`    assert construct_domain(E)[0] == ZZ[E]`

This also applies to later tests.

jksuom commented on sympy/polys/tests/test_constructor.py:
```diff
@@ -121,3 +121,18 @@ def test_precision():
+def test_issue_11538():
+    dom = ZZ[E]
+    assert construct_domain(E) == (dom, dom.convert(E))
+    dom = ZZ[x,E]
```
A space after comma. Also in several other places.

smichr commented:
If this "closes #11538" then you should add the text in quotes to your opening paragraph.

ctefer committed:
Updates from review

Removed NumberSymbol from constructor. The constructor now checks for
just is_number && is_algebraic. Updated tests to remove spaces after
comments.

jksuom commented on sympy/polys/tests/test_constructor.py:
```diff
@@ -121,3 +121,18 @@ def test_precision():
     assert y-1 > 1e-50
+
+def test_issue_11538():
+    dom = ZZ[E]
```
This `dom` line can be omitted.

jksuom commented on sympy/polys/tests/test_constructor.py:
```diff
@@ -121,3 +121,18 @@ def test_precision():
+    dom = ZZ[E]
+    assert construct_domain(E)[0] == ZZ[E]
+    dom = ZZ[x,E]
+    assert construct_domain(x**2 + 2*x + E) == \
```
This can be replaced with `assert construct_domain(x**2 + 2*x + E)[0] == ZZ[x, E]`. The `assert` statements below can be treated similarly. None of the `dom` assignments are needed and they can be omitted.

jksuom commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -1792,3 +1793,17 @@ def test_issue_2840_8155():
+
+def test_issue_11538():
+    assert solve(x + E) == [-E]
+    assert solve(x**2 + E) == [-I*sqrt(E),I*sqrt(E)]
```
Space after comma.

jksuom commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -1792,3 +1793,17 @@ def test_issue_2840_8155():
+        -cbrt(2 * E),
+        cbrt(2)*cbrt(E)/2 - cbrt(2)*sqrt(3)*I*cbrt(E)/2,
+        cbrt(2)*cbrt(E)/2 + cbrt(2)*sqrt(3)*I*cbrt(E)/2]
+    assert solve([x + 4,y + E],x,y) == {x:-4,y:-E}
```
Space after commas and colons.

jksuom replied:
This also applies to the next two `assert` lines.

ctefer replied:
Do you mean add a space or remove spaces. All of the other tests in the document contain spaces after the commas.

jksuom replied:
I mean that a space should be added after commas (and colons in dictionaries). That is a rule generally followed in SymPy code.

jksuom commented:
The failure is in the document generation unrelated to this PR.

smichr commented:
I got here by following breadcrumbs from the newest PR adding a Poly docstring. It looks like this has grown stale. I will re-submit as a new PR this evening and merge it so @jksuom can get his pending poly PRs merged.

smichr commented:
see now #12457

smichr commented:
@ctefer , I closed this since you hadn't responded for several months -- but this is not uncommon (uh hum)I should ask, however, if you are interested in finishing this, perhaps adding in a PR from myself handling the changes that have been made. If I don't hear from you within a day or so, I'll re-close this as there are other PRs waiting for the changes that you have made.
